### PR TITLE
[bugfix] Fixes #17118 Reshape tool with z support

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

As described in this ticket https://issues.qgis.org/issues/17118, the current reshape map tool resets z values. This PR fixes this issue by using `QgsPointV2` instead of `QgsPoint` for reshaping.

A simple test with a `LineStringZ` has been added.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/geometry/qgsgeometry.cpp [hunk 1]
```diff
@@ -801,27 +801,33 @@ int QgsGeometry::splitGeometry( const QList<QgsPoint>& splitLine, QList<QgsGeome
...
   reshapeLineString.setPoints( reshapeLine );
-
   QgsGeos geos( d->geometry );
   int errorCode = 0;
   QgsAbstractGeometryV2* geom = geos.reshapeGeometry( reshapeLineString, &errorCode );
+
   if ( errorCode == 0 && geom )
   {
-    detach( false );
+    detach( true );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.cpp]: Why this change? it's now cloning the geometry, only to delete it immediately after

File: src/core/geometry/qgsgeometry.cpp [hunk 2]
```diff
@@ -801,27 +801,33 @@ int QgsGeometry::splitGeometry( const QList<QgsPoint>& splitLine, QList<QgsGeome
...
     delete d->geometry;
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.cpp]: this line actually does nothing

File: src/core/geometry/qgsgeometry.cpp [hunk 3]
```diff
@@ -801,27 +801,33 @@ int QgsGeometry::splitGeometry( const QList<QgsPoint>& splitLine, QList<QgsGeome
...
-    d->geometry = geom;
     removeWkbGeos();
+    d->geometry = geom;
     return 0;
   }
+
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.cpp]: better put a "delete geom;" here - or there's potential for a leak

**Review by nyalldawson — APPROVED** [review]
Looks good - nice fix to have!

**Comment by pblottiere** [comment]
@nyalldawson You added the label `Needs Forwardporting` on this PR but I just tested on master and it seems to be working fine. Let me know if I missed something.

And thank you for cleaning/modernizing QgsGeometry :)
